### ISSUE TITLE
WPF - slider precision is not correct.

### ISSUE THREAD

**Issue by hudecekd** [dotnet/wpf#8819]
### Description

Hello,

we use slider with this configuration
```xaml
        <Slider
            Grid.Row="3"
            Grid.Column="1"
            IsSnapToTickEnabled="True"
            TickFrequency="0.1"
            Minimum="{Binding Gamma.Minimum}"
            Maximum="{Binding Gamma.Maximum}"
            Value="{Binding Gamma.Value}"/>
```
Gamma.Minimum is 0.5.
Gamma.Maximum is 4.
Gamma.Value is bounded.
All properties are `double`.
Problem is that when Value is `1.9` and i move slider very little then value does not change to `2` or `1.8` but it changes to `1.9000000000000001` which is not correct. Value should be `2`.

I have tried to debug it and I think that I have found problem at this line https://github.com/dotnet/wpf/blob/2fdb74d7b11465f36246b7caf06fa00e9c6d8454/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Controls/Slider.cs#L1135
When `* TickFrequency` is used value result of multiplication brings precision problem.

Part of the code from the link above.
```csharp
                else if (DoubleUtil.GreaterThanZero(TickFrequency))
                {
                    previous = Minimum + (Math.Round(((value - Minimum) / TickFrequency)) * TickFrequency);
                    next = Math.Min(Maximum, previous + TickFrequency);
                }
```

### Reproduction Steps

Add Slider with binding to double value

```xaml
        <Slider
            Grid.Row="3"
            Grid.Column="1"
            IsSnapToTickEnabled="True"
            TickFrequency="0.1"
            Minimum="0.5"
            Maximum="4"
            Value="{Binding Value}"/>
```

Set value to `1.9` from code.
Try to move slidre a little. See problem.

### Expected behavior

- Value should be set to `2.0`

### Actual behavior

Value is set to `1.9000000000000001`.

**Comment by lindexi** [comment]
@hudecekd Thank you for your feedback. Are you willing to fix this problem?

**Comment by wbc2048** [comment]
Hi! Do you still need help? I would love to work on this issue as a first-time contributor.

**Comment by hudecekd** [comment]
@lindexi @wbc2048 I will be glad if you try to fix it because there are many other things related to this which should be considered and I do not thing that I would be able to consider all of them.
For example: I can have Min=0.5001 and use TickFrequency=0.1 which in the end should generate `1.9001`. What I am trying to say is that it is not just about using `Math.Round` but it depends on precision on minimum and maximum also and I am not sure whether there are not other issues to consider.

So please If you know then solve it @wbc2048 . Thanks.

**Comment by mr-sergito** [comment]
Hey @wbc2048 are you working on this? If not, I'd like to give it a shot 😄

**Comment by sachPundir** [comment]
Hi, I’m currently investigating this issue and working on a potential fix.

I was able to reproduce the floating-point precision problem in SnapToTick and am testing a solution locally. If everything looks good, I’ll submit a PR.

If there are any prior discussions or preferred approaches for this fix, please let me know.

Thanks!

**Comment by sachPundir** [comment]
This fix applies rounding only to the final snapped value to eliminate floating-point artifacts (e.g., 1.9000000000000001 → 1.9) introduced by double arithmetic.

It does not round or modify Minimum, Maximum, or TickFrequency, and therefore does not destroy real precision coming from Min/Max offsets.

For example, with Minimum = 0.5001 and TickFrequency = 0.1, the snapping still correctly produces values like 1.9001.

The intent is only to remove IEEE floating-point noise, not to change logical tick precision.

**Comment by miloush** [comment]
> Problem is that when Value is 1.9 and i move slider very little then value does not change to 2 or 1.8 but it changes to 1.9000000000000001 which is not correct. Value should be 2.

Why should it be 2? The step is 0.1, it doesn't snap to the next highest value, according to the documentation it snaps to the closest value. It should snap to 1.9 for values between 1.85 and 1.95.

**Comment by hudecekd** [comment]
@miloush  You are correct. I think that my description was a little miselading because when i tried to move slider from 1.9 I expected that it ends at 1.8 or 2 when i will move it for a while but it immediatelly triggered binding change to 1.9000000000000001. So problem is that it should do "nothing" when it is 1.9 already and change to 2 or 1.8 (depending on direction of move) when appropriate (which is when it is under 1.85 or above 1.95 as you described) but since it immediatelly executes binding change it never gets chance to do that.

In ma case it caused exceptions because new value 1.9000000000000001 was not allowed in underlying model and business logic.

So
- yes it should go to 1.8 or 2 in the end
- it should not trigger binding change to 1.9000000000000001.

**Comment by sachPundir** [comment]
@miloush, I Updated the approach based on your feedback. The arbitrary rounding strategy has been replaced with decimal arithmetic for the tick index reconstruction. Math.Round(...) produces a clean integer -- the only noise was in multiplying that integer back by TickFrequency in double. Performing that final multiplication in decimal eliminates the artifact without introducing any new rounding logic. A try/catch fallback handles values outside decimal range. Happy to adjust if you have further concerns.

**Comment by sachPundir** [comment]
Updated PR #11431 with a revised approach following reviewer feedback. The fix now uses decimal arithmetic to eliminate IEEE 754 noise in the tick reconstruction step, replacing the earlier rounding approach.

**Comment by sachPundir** [comment]
After exploring multiple approaches in PR #11431, I wanted to share a conclusion from the investigation.

The observed behavior appears to be a natural result of IEEE floating-point arithmetic rather than an issue in the Slider snapping logic itself. While artifacts like 1.9000000000000001 can be surprising, they are consistent with how floating-point values behave.

Attempts to “fix” this at the framework level (e.g., rounding or decimal reconstruction) introduce trade-offs such as performance overhead, added complexity, or incorrect results for certain tick magnitudes.

Based on this, it seems more appropriate for consumers (e.g., UI display or binding layers) to handle formatting if needed, rather than modifying the core snapping logic.

Sharing this in case it helps with deciding how to proceed with the issue.
